### ISSUE TITLE
istioctl analyze fails to parse some non-istio yaml files

### ISSUE THREAD

**Issue by selmanj** [istio/istio#19336]
**Bug description**

I discovered this when setting up a GitOps-style workflow for a cluster. Running `istioctl x analyze` against the manifest generated by `istioctl manifest` produces a lot of parse errors (most seem to be against objects without a 'spec' field).

Here's a command that produces what I observed:
```
jsselman-imacpro:~/s/example-istio-cluster (master|●1) [65] $ ~/go/out/darwin_amd64/release/istioctl  x analyze /tmp/istio-1.4.0.yaml 
Error(s) reading files: 1 error occurred:
        * errors parsing content "/tmp/istio-1.4.0.yaml": 70 errors occurred:
        * error processing /tmp/istio-1.4.0.yaml[0]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRole
        * error processing /tmp/istio-1.4.0.yaml[1]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRoleBinding
        * error processing /tmp/istio-1.4.0.yaml[2]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[3]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[4]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[5]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[6]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[7]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[8]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[9]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[10]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[11]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[12]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[13]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[14]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[15]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[16]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[17]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[18]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[19]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[20]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[21]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[22]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[23]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[24]: failed finding spec for group/kind: apiextensions.k8s.io/CustomResourceDefinition
        * error processing /tmp/istio-1.4.0.yaml[26]: failed finding spec for group/kind: /ServiceAccount
        * error processing /tmp/istio-1.4.0.yaml[27]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRole
        * error processing /tmp/istio-1.4.0.yaml[28]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRoleBinding
        * error processing /tmp/istio-1.4.0.yaml[30]: failed finding spec for group/kind: policy/PodDisruptionBudget
        * error processing /tmp/istio-1.4.0.yaml[32]: failed finding spec for group/kind: /ServiceAccount
        * error processing /tmp/istio-1.4.0.yaml[33]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRole
        * error processing /tmp/istio-1.4.0.yaml[34]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRoleBinding
        * error processing /tmp/istio-1.4.0.yaml[35]: failed finding spec for group/kind: /ConfigMap
        * error processing /tmp/istio-1.4.0.yaml[36]: failed finding spec for group/kind: /ConfigMap
        * error processing /tmp/istio-1.4.0.yaml[37]: failed finding spec for group/kind: /ConfigMap
        * error processing /tmp/istio-1.4.0.yaml[39]: failed finding spec for group/kind: policy/PodDisruptionBudget
        * error processing /tmp/istio-1.4.0.yaml[41]: failed finding spec for group/kind: /ServiceAccount
        * error processing /tmp/istio-1.4.0.yaml[42]: failed finding spec for group/kind: autoscaling/HorizontalPodAutoscaler
        * error processing /tmp/istio-1.4.0.yaml[45]: failed finding spec for group/kind: policy/PodDisruptionBudget
        * error processing /tmp/istio-1.4.0.yaml[47]: failed finding spec for group/kind: /ServiceAccount
        * error processing /tmp/istio-1.4.0.yaml[49]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRole
        * error processing /tmp/istio-1.4.0.yaml[50]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRoleBinding
        * error processing /tmp/istio-1.4.0.yaml[51]: failed finding spec for group/kind: /ConfigMap
        * error processing /tmp/istio-1.4.0.yaml[53]: failed finding spec for group/kind: admissionregistration.k8s.io/MutatingWebhookConfiguration
        * error processing /tmp/istio-1.4.0.yaml[54]: failed finding spec for group/kind: policy/PodDisruptionBudget
        * error processing /tmp/istio-1.4.0.yaml[56]: failed finding spec for group/kind: /ServiceAccount
        * error processing /tmp/istio-1.4.0.yaml[57]: failed finding spec for group/kind: /ConfigMap
        * error processing /tmp/istio-1.4.0.yaml[58]: failed finding spec for group/kind: autoscaling/HorizontalPodAutoscaler
        * error processing /tmp/istio-1.4.0.yaml[59]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRole
        * error processing /tmp/istio-1.4.0.yaml[60]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRoleBinding
        * error processing /tmp/istio-1.4.0.yaml[61]: failed finding spec for group/kind: /ConfigMap
        * error processing /tmp/istio-1.4.0.yaml[62]: failed finding spec for group/kind: /ConfigMap
        * error processing /tmp/istio-1.4.0.yaml[65]: failed finding spec for group/kind: policy/PodDisruptionBudget
        * error processing /tmp/istio-1.4.0.yaml[67]: failed finding spec for group/kind: /ServiceAccount
        * error processing /tmp/istio-1.4.0.yaml[68]: failed finding spec for group/kind: autoscaling/HorizontalPodAutoscaler
        * error processing /tmp/istio-1.4.0.yaml[69]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRole
        * error processing /tmp/istio-1.4.0.yaml[70]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRoleBinding
        * error processing /tmp/istio-1.4.0.yaml[73]: failed finding spec for group/kind: policy/PodDisruptionBudget
        * error processing /tmp/istio-1.4.0.yaml[75]: failed finding spec for group/kind: /ServiceAccount
        * error processing /tmp/istio-1.4.0.yaml[76]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRole
        * error processing /tmp/istio-1.4.0.yaml[77]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRoleBinding
        * error processing /tmp/istio-1.4.0.yaml[78]: failed finding spec for group/kind: /ConfigMap
        * error processing /tmp/istio-1.4.0.yaml[81]: failed finding spec for group/kind: /ServiceAccount
        * error processing /tmp/istio-1.4.0.yaml[82]: failed finding spec for group/kind: autoscaling/HorizontalPodAutoscaler
        * error processing /tmp/istio-1.4.0.yaml[83]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRole
        * error processing /tmp/istio-1.4.0.yaml[84]: failed finding spec for group/kind: rbac.authorization.k8s.io/ClusterRoleBinding
        * error processing /tmp/istio-1.4.0.yaml[105]: failed finding spec for group/kind: /ConfigMap
        * error processing /tmp/istio-1.4.0.yaml[107]: failed finding spec for group/kind: policy/PodDisruptionBudget
        * error processing /tmp/istio-1.4.0.yaml[109]: failed finding spec for group/kind: /ServiceAccount
        * error processing /tmp/istio-1.4.0.yaml[110]: failed finding spec for group/kind: /



No validation issues found (but 1 file could not be parsed)
Error: Some files couldn't be parsed.
```

**Expected behavior**

The command 
```
~/src/istio-1.4.0/bin/istioctl manifest generate > /tmp/istio-1.4.0.yaml && ~/src/istio-1.4.0/bin/istioctl experimental analyze /tmp/istio-1.4.0.yaml
```
should emit no errors.

**Steps to reproduce the bug**

Run:
```
~/src/istio-1.4.0/bin/istioctl manifest generate > /tmp/istio-1.4.0.yaml && ~/src/istio-1.4.0/bin/istioctl experimental analyze /tmp/istio-1.4.0.yaml
```
and see a lot of errors.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

1.4.0 (also fails on master).

**Comment by sushicw** [comment]
To clarify what's happening: the issue here isn't a lack of "spec" field, it's that the group/kind isn't recognized by the code doing the parsing. Instances of the custom resources are recognized, as are some K8s types (e.g. Namespace, Deployment, Pod, Service) but other types are not recognized, and thus can't be converted. We only recognize what's in `metadata.yaml`.

This doesn't come up in Galley because we're not listening to those resources, so Galley never sees them. By having them in a yaml file, we sort of force the issue. 

I think skipping over resources like this is the correct behavior, but maybe we're being too aggressive in how we report the parse errors in this case. 

Maybe we could include a list (probably in metadata.yaml) of "types we recognize but don't process" and skip over them without generating an error.

**Comment by selmanj** [comment]
I strongly prefer we skip over resources we don't recognize, rather than trying to build an explicit whitelist.

**Comment by sushicw** [comment]
We do skip over them, the question is how noisy to be about it. :) Put another way: do we need to distinguish between "this resource is mistyped, thus I couldn't find it" and "this resource isn't one I care about, thus I couldn't find it but don't care" ?

For the `istioctl analyze` case, I think we're probably OK skipping over unrecognized resource errors, maybe with some extra output in `--verbose` to indicate that we did so but without printing errors or forcing a nonzero exit. (Other types of parse errors should, of course, continue to break).

**Comment by selmanj** [comment]
Ah, got it. Thanks for explaining - I agree that an "unknown type" error should be ignored when using istioctl analyze (and other errors should cause breakage).

**Comment by selmanj** [comment]
This should be resolved as of istio 1.5 (we ignore any errors for types we don't know or care about).
